flask/app/variants.py

```diff
@@ -228,17 +227,151 @@ def parse_gene_panel() -> List[Any]:
         app.logger.error("Not all requested genes were found.")
         abort(400, description="Not all requested genes were found.")
 
-    genes = query.all()
+    # genes_list = query.all()
+
+    app.logger.debug("Found %d genes with panel %s", found_genes, ensgs)
+
+    return gene_filter
+
+
+def parse_region(region: str):
+    """
+    Parses a given list of ranges (eg. "chr1:0500-0509,chr5:0600-0630,chr18:0440-0330").
+    Returns a filter for use against the variant table to find corresponding variants
+    based on gene start and end positions.
+    """
+    CHR, START, END = 0, 1, 2
+    # format check
+    # 1: chromosome, 2: start position, 3: end position
+    region_pattern = re.compile(r"^chr([\da-zA-Z]+):([\d]+)-([\d]+)$")
+    matches = [region_pattern.match(r) for r in region.split(",")]
+
+    app.logger.info("Requested region: %s", region)
+
+    if not all(match is not None and len(match.groups()) == 3 for match in matches):
+        app.logger.error("Invalid region format: %s", region)
+        abort(400, description="Invalid region format")
+
+    # prepare set of regions
+    region_list = []
+    for r in set([match.groups() for match in matches]):
+        # fix regions to be consistent (start-end)
+        if int(r[START]) > int(r[END]):
+            region_list.append((r[CHR], r[END], r[START]))
+        else:
+            region_list.append(r)
+
+    # query
+    region_filter = or_(
+        *[
+            and_(
+                models.Variant.chromosome == tup[CHR],
+                int(tup[START]) <= models.Variant.position,
+                models.Variant.position <= int(tup[END]),
+            )
+            for tup in region_list
+        ]
+    )
+
+    return region_filter
+
+
+def parse_position(position: str):
+    """
+    Parses a given list of positions and chromosomes (eg. "chr1:5000,chrY:8000").
+    Returns a filter for use against the variant table to find corresponding variants
+    based on gene start and end positions.
+    """
+    CHR, POS = 0, 1
+    # format check
+    # 1: chromosome, 2: position
+    position_pattern = re.compile(r"^chr([\da-zA-Z]+):([\d]+)$")
+    matches = [position_pattern.match(p) for p in position.split(",")]
+
+    app.logger.info("Requested position: %s", position)
+
+    if not all(match is not None and len(match.groups()) == 2 for match in matches):
+        app.logger.error("Invalid position format: %s", position)
+        abort(400, description="Invalid position format")
+
+    # prepare set of positions
+    position_set = set([match.groups() for match in matches])
+
+    # query
+    position_filter = or_(
+        *[
+            and_(
+                models.Variant.chromosome == tup[CHR].upper(),
+                models.Variant.position == tup[POS],
+            )
+            for tup in position_set
+        ]
+    )
+
+    return position_filter
+
+
+def parse_rsid(rsid: str):
```

Comment:
Not an issue with the PR itself but unfortunately a query with a single rsid tied to >1000 variants takes over 10 minutes to execute for a ptp-wise report since it's a text search. I don't see a way around this without decomposing the rsid column into a 1:n table. At the same time this does bring us back to the problem where we might be re-inventing the creation of the reports.